crates/bifrost/src/loglets/local_loglet/log_store.rs

```diff
@@ -0,0 +1,217 @@
+use tracing::{debug, warn};
+
+use super::keys::{MetadataKey, MetadataKind};
+use super::log_state::{log_state_full_merge, log_state_partial_merge, LogState};
+use super::log_store_writer::{RocksDbLogWriter, WriterOptions};
+use super::Options;
+
+pub(crate) static DATA_CF: &str = "logstore_data";
+pub(crate) static METADATA_CF: &str = "logstore_metadata";
+
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum LogStoreError {
+    #[error(transparent)]
+    // unfortunately, we have to use Arc here, because the bincode error is not Clone.
+    Encode(#[from] Arc<bincode::error::EncodeError>),
+    #[error(transparent)]
+    // unfortunately, we have to use Arc here, because the bincode error is not Clone.
+    Decode(#[from] Arc<bincode::error::DecodeError>),
+    #[error(transparent)]
+    Rocksdb(#[from] rocksdb::Error),
+}
+
+#[derive(Debug, Clone)]
+pub struct RocksDbLogStore {
+    db: Arc<DB>,
+}
+
+impl RocksDbLogStore {
+    pub fn new(storage_path: &Path, options: &Options) -> anyhow::Result<Self> {
+        let cache = if options.rocksdb_cache_size > 0 {
+            Some(Cache::new_lru_cache(options.rocksdb_cache_size))
+        } else {
+            None
+        };
+
+        let cfs = [
+            rocksdb::ColumnFamilyDescriptor::new(DATA_CF, cf_options(options, cache.clone())),
+            rocksdb::ColumnFamilyDescriptor::new(METADATA_CF, cf_options(options, cache.clone())),
+        ];
+        let db_options = db_options(options);
+
+        let db = DB::open_cf_descriptors(&db_options, storage_path, cfs)
+            .context("failed to open rocksdb for local loglet store")?;
+
+        Ok(Self { db: Arc::new(db) })
+    }
+
+    pub fn data_cf(&self) -> &rocksdb::ColumnFamily {
+        self.db.cf_handle(DATA_CF).expect("DATA_CF exists")
+    }
+
+    pub fn metadata_cf(&self) -> &rocksdb::ColumnFamily {
+        self.db.cf_handle(METADATA_CF).expect("METADATA_CF exists")
+    }
+
+    pub fn get_log_state(&self, log_id: u64) -> Result<Option<LogState>, LogStoreError> {
+        let metadata_cf = self.metadata_cf();
+        let value = self.db.get_pinned_cf(
+            metadata_cf,
+            MetadataKey::new(log_id, MetadataKind::LogState).to_bytes(),
+        )?;
+
+        if let Some(value) = value {
+            Ok(Some(LogState::from_slice(&value)?))
+        } else {
+            Ok(None)
+        }
+    }
+
+    pub fn create_writer(&self, opts: WriterOptions) -> RocksDbLogWriter {
+        RocksDbLogWriter::new(self.db.clone(), opts)
+    }
+
+    pub fn db(&self) -> &DB {
+        &self.db
+    }
+
+    pub fn shutdown(&self) {
+        let start = Instant::now();
+        if let Err(e) = self.db.flush_wal(true) {
+            warn!("Failed to flush local loglet rocksdb WAL: {}", e);
+        }
+        self.db.cancel_all_background_work(true);
+        debug!(
+            "Local loglet clean rocksdb shutdown took {:?}",
+            start.elapsed(),
+        );
+    }
```

Comment:
This is nice :-) Should we do something similar for the graceful shutdown of the PP state storage?

Reply:
Good call. I think potentially yes. I'm being extra defensive for the logstore part, not sure if this will bring immediate value to PP state storage at the moment though.